File: ClipboardZanager/Tests/ClipboardZanager.Shared.Tests/CloudStorage/CloudStorageTests.cs
using System;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using ClipboardZanager.Shared.Exceptions;
using ClipboardZanager.Shared.Tests.Mocks;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace ClipboardZanager.Shared.Tests.CloudStorage
{
    [TestClass]
    public class CloudStorageTests
    {
        [TestMethod]
        public async Task CloudAuthentication()
        {
            var authentication = new CloudAuthenticationMock();

            var success = await authentication.AuthenticateAsync("http://authenticationUri", "http://exceptedUi");
            var fail = await authentication.AuthenticateAsync(string.Empty, "http://exceptedUi");

            Assert.IsFalse(success.IsCanceled);
            Assert.AreEqual(success.RedirectedUri.OriginalString, "http://exceptedUi");
            Assert.IsTrue(fail.IsCanceled);
        }

        [TestMethod]
        public void CloudTokenProvider()
        {
            var tokenProvider = new CloudTokenProviderMock();

            var token = tokenProvider.GetToken("MyToken");
            Assert.AreEqual(token, "123456789abc");

            try
            {
                var token2 = tokenProvider.GetToken("MyToken2");
                Assert.Fail();
            }
            catch
            {
            }
        }

        [TestMethod]
        public async Task CloudStorageProvider()
        {
            var storageProvider = new CloudStorageProviderMock();
            var stream = new MemoryStream();

            Assert.AreEqual(storageProvider.CloudServiceName, "MockCloudStorageProvider");
            Assert.IsFalse(storageProvider.IsAuthenticated);
            Assert.IsTrue(storageProvider.CredentialExists);

            try
            {
                await storageProvider.DownloadFileAsync("/path/MyApp/myFile.txt", stream);
                Assert.Fail();
            }
            catch (NotAuthenticatedException)
            {
            }
            catch (Exception)
            {
                Assert.Fail();
            }

            Assert.IsFalse(await storageProvider.TryAuthenticateAsync());
            Assert.IsFalse(storageProvider.IsAuthenticated);

            Assert.IsTrue(await storageProvider.TryAuthenticateWithUiAsync(new CloudAuthenticationMock()));
            Assert.IsTrue(storageProvider.IsAuthenticated);

            try
            {
                await storageProvider.DownloadFileAsync("/path/MyApp/myFile.txt", null);
                Assert.Fail();
            }
            catch (NullReferenceException)
            {
            }
            catch (Exception)
            {
                Assert.Fail();
            }

            try
            {
                await storageProvider.DownloadFileAsync("/path/MyApp/myFile2.txt", stream);
                Assert.Fail();
            }
            catch (FileNotFoundException)
            {
            }
            catch (Exception)
            {
                Assert.Fail();
            }

            await storageProvider.DownloadFileAsync("/path/MyApp/myFile.txt", stream);
            Assert.IsTrue(stream.ToArray().SequenceEqual(new byte[] { 0, 1, 2, 3, 4 }));

            Assert.IsTrue(storageProvider.IsAuthenticated);
            await storageProvider.SignOutAsync();
            Assert.IsFalse(storageProvider.IsAuthenticated);
        }
    }
}

File: Android/app/src/test/java/com/etiennebaudoux/clipboardzanager/componentmodel/cloudstorage/CloudStorageTests.java
package com.etiennebaudoux.clipboardzanager.componentmodel.cloudstorage;

import com.etiennebaudoux.clipboardzanager.componentmodel.exceptions.NotAuthenticatedException;
import com.etiennebaudoux.clipboardzanager.mocks.CloudAuthenticationMock;
import com.etiennebaudoux.clipboardzanager.mocks.CloudStorageProviderMock;
import com.etiennebaudoux.clipboardzanager.mocks.CloudTokenProviderMock;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.io.ByteArrayOutputStream;
import java.io.FileNotFoundException;
import java.util.Arrays;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

@RunWith(RobolectricTestRunner.class)
public class CloudStorageTests {
    @Test
    public void cloudAuthentication() {
        CloudAuthenticationMock authentication = new CloudAuthenticationMock();

        AuthenticationResult success = authentication.authenticateAsync("http://authenticationUri", "http://exceptedUi").await();
        AuthenticationResult fail = authentication.authenticateAsync("", "http://exceptedUi").await();

        assertFalse(success.isCanceled());
        assertEquals("http://exceptedUi", success.getRedirectedUri());
        assertTrue(fail.isCanceled());
    }

    @Test
    public void cloudTokenProvider() {
        CloudTokenProviderMock tokenProvider = new CloudTokenProviderMock();

        String token = tokenProvider.getToken("MyToken");
        assertEquals("123456789abc", token);

        try {
            String token2 = tokenProvider.getToken("MyToken2");
            fail();
        } catch (Exception ex) {
        }
    }

    @Test
    public void cloudStorageProvider() {
        CloudStorageProviderMock storageProvider = new CloudStorageProviderMock();
        ByteArrayOutputStream stream = new ByteArrayOutputStream();

        assertEquals("MockCloudStorageProvider", storageProvider.getCloudServiceName());
        assertFalse(storageProvider.isAuthenticated());
        assertTrue(storageProvider.credentialExists());

        try {
            storageProvider.downloadFileAsync("/path/MyApp/myFile.txt", stream).await();
            fail();
        } catch (RuntimeException ex) {
            assertTrue(ex.getCause() instanceof NotAuthenticatedException);
        } catch (Exception ex) {
            fail();
        }

        assertFalse(storageProvider.tryAuthenticateAsync().await());
        assertFalse(storageProvider.isAuthenticated());

        assertTrue(storageProvider.tryAuthenticateWithUiAsync(true).await());
        assertTrue(storageProvider.isAuthenticated());

        try {
            storageProvider.downloadFileAsync("/path/MyApp/myFile.txt", null).await();
            fail();
        } catch (RuntimeException ex) {
            assertTrue(ex.getCause() instanceof NullPointerException);
        } catch (Exception ex) {
            fail();
        }

        try {
            storageProvider.downloadFileAsync("/path/MyApp/myFile2.txt", stream).await();
            fail();
        } catch (RuntimeException ex) {
            assertTrue(ex.getCause() instanceof FileNotFoundException);
        } catch (Exception ex) {
            fail();
        }

        storageProvider.downloadFileAsync("/path/MyApp/myFile.txt", stream).await();
        assertTrue(Arrays.equals(stream.toByteArray(), new byte[] { 0, 1, 2, 3, 4 }));

        assertTrue(storageProvider.isAuthenticated());
        storageProvider.signOutAsync().await();
        assertFalse(storageProvider.isAuthenticated());
    }
}
